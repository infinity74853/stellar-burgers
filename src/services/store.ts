import { configureStore, combineReducers } from '@reduxjs/toolkit';
import {
  TypedUseSelectorHook,
  useDispatch as dispatchHook,
  useSelector as selectorHook
} from 'react-redux';
import ingredientsReducer from './slices/ingredientsSlice';
import userReducer from './slices/userSlice';
import feedReducer from './slices/feedSlice';
import ordersReducer from './slices/ordersSlice';
import burgerConstructorReducer from './slices/burgerConstructorSlice';

const rootReducer = combineReducers({
  ingredients: ingredientsReducer,
  orders: ordersReducer,
  user: userReducer,
  feed: feedReducer,
  burgerConstructor: burgerConstructorReducer
});

export const store = configureStore({
  reducer: rootReducer,
  devTools: process.env.NODE_ENV !== 'production'
});

export type RootState = ReturnType<typeof rootReducer>;
export type AppDispatch = typeof store.dispatch;

export const useDispatch: () => AppDispatch = () => dispatchHook();
export const useSelector: TypedUseSelectorHook<RootState> = selectorHook;

export default store;
