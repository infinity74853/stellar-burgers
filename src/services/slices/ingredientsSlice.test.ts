import ingredientsReducer, {
  fetchIngredients,
  fetchIngredientById,
  setCurrentIngredient,
  clearCurrentIngredient,
  selectIngredients,
  selectIngredientsLoading,
  selectIngredientsError,
  selectIngredientsLoaded,
  selectCurrentIngredient
} from './ingredientsSlice';
import { RootState } from '../store';
import { TIngredient } from '@utils-types';

// Создаём копию initialState из слайса (без его изменения)
const initialState = {
  ingredients: [],
  isLoading: false,
  error: null,
  loaded: false,
  currentIngredient: null
};

// Пример мокового ингредиента
const mockIngredient: TIngredient = {
  _id: '1',
  name: 'Булка',
  type: 'bun',
  proteins: 10,
  fat: 20,
  carbohydrates: 30,
  calories: 40,
  price: 100,
  image: 'image-url',
  image_mobile: 'mobile-image-url',
  image_large: 'large-image-url'
};

const mockIngredients: TIngredient[] = [
  mockIngredient,
  {
    ...mockIngredient,
    _id: '2',
    name: 'Соус',
    type: 'sauce'
  }
];

describe('Тесты для ingredientsSlice', () => {
  describe('Начальное состояние', () => {
    it('должно соответствовать ожидаемой структуре', () => {
      const state = ingredientsReducer(undefined, { type: '' });
      expect(state).toEqual(initialState);
    });

    it('совпадает со снапшотом', () => {
      expect(ingredientsReducer(undefined, { type: '' })).toMatchSnapshot();
    });
  });

  describe('Синхронные экшены', () => {
    it('setCurrentIngredient устанавливает ингредиент', () => {
      const state = ingredientsReducer(
        initialState,
        setCurrentIngredient(mockIngredient)
      );
      expect(state.currentIngredient).toEqual(mockIngredient);
      expect(state).toMatchSnapshot();
    });

    it('clearCurrentIngredient очищает текущий ингредиент', () => {
      const stateWithIngredient = {
        ...initialState,
        currentIngredient: mockIngredient
      };
      const state = ingredientsReducer(
        stateWithIngredient,
        clearCurrentIngredient()
      );
      expect(state.currentIngredient).toBeNull();
      expect(state).toMatchSnapshot();
    });
  });

  describe('Асинхронные экшены', () => {
    it('fetchIngredients.pending устанавливает флаг загрузки', () => {
      const state = ingredientsReducer(initialState, {
        type: fetchIngredients.pending.type
      });
      expect(state.isLoading).toBe(true);
      expect(state.error).toBeNull();
      expect(state).toMatchSnapshot();
    });

    it('fetchIngredients.fulfilled сохраняет ингредиенты', () => {
      const action = {
        type: fetchIngredients.fulfilled.type,
        payload: mockIngredients
      };
      const state = ingredientsReducer(initialState, action);
      expect(state.ingredients).toEqual(mockIngredients);
      expect(state.isLoading).toBe(false);
      expect(state.loaded).toBe(true);
      expect(state).toMatchSnapshot();
    });

    it('fetchIngredients.rejected сохраняет ошибку', () => {
      const error = 'Ошибка загрузки';
      const action = { type: fetchIngredients.rejected.type, payload: error };
      const state = ingredientsReducer(initialState, action);
      expect(state.error).toBe(error);
      expect(state.isLoading).toBe(false);
      expect(state).toMatchSnapshot();
    });
  });

  describe('Селекторы', () => {
    const testState: RootState = {
      ingredients: {
        ...initialState,
        ingredients: mockIngredients,
        currentIngredient: mockIngredient
      },
      // Заглушки для других слайсов
      burgerConstructor: {} as any,
      feed: {} as any,
      orderHistory: {} as any,
      user: {} as any
    };

    it('selectIngredients возвращает ингредиенты', () => {
      expect(selectIngredients(testState)).toEqual(mockIngredients);
    });

    it('selectCurrentIngredient возвращает текущий ингредиент', () => {
      expect(selectCurrentIngredient(testState)).toEqual(mockIngredient);
    });

    it('комбинация селекторов совпадает со снапшотом', () => {
      expect({
        ingredients: selectIngredients(testState),
        current: selectCurrentIngredient(testState),
        loading: selectIngredientsLoading(testState),
        error: selectIngredientsError(testState),
        loaded: selectIngredientsLoaded(testState)
      }).toMatchSnapshot();
    });
  });
});
