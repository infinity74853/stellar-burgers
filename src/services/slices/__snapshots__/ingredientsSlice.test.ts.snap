// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тесты для ingredientsSlice Асинхронные экшены fetchIngredients.fulfilled сохраняет ингредиенты 1`] = `
{
  "currentIngredient": null,
  "error": null,
  "ingredients": [
    {
      "_id": "1",
      "calories": 40,
      "carbohydrates": 30,
      "fat": 20,
      "image": "image-url",
      "image_large": "large-image-url",
      "image_mobile": "mobile-image-url",
      "name": "Булка",
      "price": 100,
      "proteins": 10,
      "type": "bun",
    },
    {
      "_id": "2",
      "calories": 40,
      "carbohydrates": 30,
      "fat": 20,
      "image": "image-url",
      "image_large": "large-image-url",
      "image_mobile": "mobile-image-url",
      "name": "Соус",
      "price": 100,
      "proteins": 10,
      "type": "sauce",
    },
  ],
  "isLoading": false,
  "loaded": true,
}
`;

exports[`Тесты для ingredientsSlice Асинхронные экшены fetchIngredients.pending устанавливает флаг загрузки 1`] = `
{
  "currentIngredient": null,
  "error": null,
  "ingredients": [],
  "isLoading": true,
  "loaded": false,
}
`;

exports[`Тесты для ingredientsSlice Асинхронные экшены fetchIngredients.rejected сохраняет ошибку 1`] = `
{
  "currentIngredient": null,
  "error": "Ошибка загрузки",
  "ingredients": [],
  "isLoading": false,
  "loaded": false,
}
`;

exports[`Тесты для ingredientsSlice Начальное состояние совпадает со снапшотом 1`] = `
{
  "currentIngredient": null,
  "error": null,
  "ingredients": [],
  "isLoading": false,
  "loaded": false,
}
`;

exports[`Тесты для ingredientsSlice Селекторы комбинация селекторов совпадает со снапшотом 1`] = `
{
  "current": {
    "_id": "1",
    "calories": 40,
    "carbohydrates": 30,
    "fat": 20,
    "image": "image-url",
    "image_large": "large-image-url",
    "image_mobile": "mobile-image-url",
    "name": "Булка",
    "price": 100,
    "proteins": 10,
    "type": "bun",
  },
  "error": null,
  "ingredients": [
    {
      "_id": "1",
      "calories": 40,
      "carbohydrates": 30,
      "fat": 20,
      "image": "image-url",
      "image_large": "large-image-url",
      "image_mobile": "mobile-image-url",
      "name": "Булка",
      "price": 100,
      "proteins": 10,
      "type": "bun",
    },
    {
      "_id": "2",
      "calories": 40,
      "carbohydrates": 30,
      "fat": 20,
      "image": "image-url",
      "image_large": "large-image-url",
      "image_mobile": "mobile-image-url",
      "name": "Соус",
      "price": 100,
      "proteins": 10,
      "type": "sauce",
    },
  ],
  "loaded": false,
  "loading": false,
}
`;

exports[`Тесты для ingredientsSlice Синхронные экшены clearCurrentIngredient очищает текущий ингредиент 1`] = `
{
  "currentIngredient": null,
  "error": null,
  "ingredients": [],
  "isLoading": false,
  "loaded": false,
}
`;

exports[`Тесты для ingredientsSlice Синхронные экшены setCurrentIngredient устанавливает ингредиент 1`] = `
{
  "currentIngredient": {
    "_id": "1",
    "calories": 40,
    "carbohydrates": 30,
    "fat": 20,
    "image": "image-url",
    "image_large": "large-image-url",
    "image_mobile": "mobile-image-url",
    "name": "Булка",
    "price": 100,
    "proteins": 10,
    "type": "bun",
  },
  "error": null,
  "ingredients": [],
  "isLoading": false,
  "loaded": false,
}
`;
